sys/net/application_layer/nanocoap/sock.c

```diff
@@ -139,13 +158,16 @@ ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
             /* parse response */
             if (coap_parse(pkt, payload, res) < 0) {
                 DEBUG("nanocoap: error parsing packet\n");
-                res = -EBADMSG;
+                retry_rx = true;
+                continue;
             }
             else if (coap_get_id(pkt) != id) {
```

Comment:
Ouch, that is not correct. CoAP has the message ID and the Token, and they are often confused with each other - as done here.

The Message ID is for de-duplication only. E.g. a CON message will result in an ACK or RST message that matches the ID, and in either case the retransmissions have to stop. If a receiver gets the same message ID a second time within some time frame from the same endpoint received at the same endpoint, it is a duplicate. Servers that host resources on which non-idempotent operations can be performed on (e.g. a `POST` to `/speaker/volume/increase`) must filter out duplicates, other servers can just handle duplicates again.

The CoAP Token is used to match requests and responses. E.g. a CON request with ID 1337 and Token 0xacdc could get the reply in an ACK with ID 1337 and Token 0xacdc. But it could also receive an empty ACK with ID 1337 (empty ACKs and RSTs have no Token, which is not to be confused with a zero-length Token). And the reply would then be send as separate response either as CON or NON with a server chosen message id (e.g. ID 42), but the Token would have to be 0xacdc so that the client can match it to the request.

So, the implementation here is fundamentally broken.

Reply:
I think this was already the case before since this was inherited from the suit located implementation: https://github.com/benpicco/RIOT/blob/0d8070fc429aae020f75873184312656b4cf3861/sys/suit/transport/coap.c#L185-L188

So the check there should actually be against the token?

Reply:
So 

```C
if ((coap_get_type(pkt) == COAP_TYPE_RST) || (coap_get_type(pkt) == COAP_TYPE_ACK) {
    if (coap_get_id(pkt) != id) {
        continue;
    }
} else {
    /* TODO: check token */
}
```

Reply:
8eb5a67e89e90fa61bd02850ba487b6a49dbded4